docs(api): Jupyter modules workaround and edits to Advanced Control page (#11510)
OpenTrons/opentrons-api: api/docs/v2/new_advanced_running.rst

----------------

The OT-2 runs a `Jupyter Notebook <https://jupyter.org>`_ server on port 48888, which you can connect to with your web browser. This is a convenient environment for writing and debugging protocols, since you can define different parts of your protocol in different notebook cells and run a single cell at a time.

To access the OT-2’s Jupyter Notebook, either:

- go to the **Advanced** tab of Robot Settings and click the **Launch Jupyter Notebook** button, or
- if you know your robot's IP address, navigate directly to ``http://<robot-ip>:48888`` in your web browser.

Once you've launched Jupyter Notebook, you can create a notebook file or edit an existing one. These notebook files are stored on the OT-2 itself. If you want to save code from a notebook to your computer, go to **File > Download As** in the notebook interface.

Protocol Structure
++++++++++++++++++

Jupyter Notebook is structured around `cells`: discrete chunks of code that can be run individually. This is nearly the opposite of Opentrons protocols, which bundle all commands into a single ``run`` function. Therefore, to take full advantage of Jupyter Notebook, you have to restructure your protocol. 

> Rather than writing a  ``run`` function and embedding commands within it, start your notebook with ``import opentrons.execute`` and an instantiation of  :py:meth:`opentrons.execute.get_protocol_api`. This method also replaces the metadata block of a standalone protocol by taking the minimum :ref:`API version <v2-versioning>` as its argument. Then you can call :py:class:`~opentrons.protocol_api.contexts.ProtocolContext` methods in subsequent lines or cells:

Reviewer: Terminology pedantry:

* We call the `opentrons.execute.get_protocol_api
()` function, not instantiate it.
* `opentrons.execute.get_protocol_api()` is a function, not a method.

```suggestion
Rather than writing a  ``run`` function and embedding commands within it, start your notebook by importing ``opentrons.execute`` and calling :py:meth:`opentrons.execute.get_protocol_api`. This function also replaces the ``metadata`` block of a standalone protocol by taking the minimum :ref:`API version <v2-versioning>` as its argument. Then you can call :py:class:`~opentrons.protocol_api.contexts.ProtocolContext` methods in subsequent lines or cells:
```

Author: Pedantry always appreciated! The API reference page clearly tags properties but I wish it were better about functions vs. methods.

    protocol = opentrons.execute.get_protocol_api('|apiLevel|')

>     protocol.home()

Reviewer: Would it be helpful to add real protocol stuff in here, like loading a pipette and some labware and doing a transfer?

Author: A full Jupyter tutorial is in the backlog — that would definitely be the place to show this in greater detail.

The first command you execute should always be :py:meth:`~opentrons.protocol_api.contexts.ProtocolContext.home`. If you try to execute other commands first, you will get a ``MustHomeError``. (When running protocols through the Opentrons App, the robot homes automatically.)

> All subsequent method calls will use the same :py:class:`.ProtocolContext`, unless you call :py:meth:`~opentrons.execute.get_protocol_api` again. If you do, the robot will update its cache of attached instruments and modules. Calling :py:meth:`~opentrons.execute.get_protocol_api` repeatedly will return an entirely new :py:class:`.ProtocolContext` each time, without loading any labware or instruments. This is a good way to reset the state of the system if you loaded something incorrectly or want to start from the beginning of your protocol logic.

Reviewer: I think it's kind of confusing to say "All subsequent method calls will use the same `ProtocolContext`" here.

Subsequent method calls will use whichever `ProtocolContext` the user is calling them on. In other words, there's no global "current `ProtocolContext`" variable that the methods will automatically use.

Author: I'm mostly passing this along from previous versions of the doc, but I think what they're getting at is that as you run cells in Jupyter and call `protocol.whatever_method()` it keeps mutating `protocol`. For example, if you have a cell that loads labware in slot 2, the first time you run it, the labware will load; if you run it again, you'll get an error (the labware is already loaded). But if you run your cell with `get_protocol_api` again, it will "reset" `protocol`.

So maybe the change here is to say: 
```
All subsequent method calls will update the state of your :py:class:`.ProtocolContext`, 
unless you reset it by calling :py:meth:`~opentrons.execute.get_protocol_api` again.
```
…and perhaps remove the sentence beginning `If you do`.

Reviewer: > I think what they're getting at is that as you run cells in Jupyter and call protocol.whatever_method() it keeps mutating protocol. For example, if you have a cell that loads labware in slot 2, the first time you run it, the labware will load; if you run it again, you'll get an error (the labware is already loaded).

Right, exactly, but this isn't something specific to Jupyter. This is just how the Python Protocol API, and Python code in general, works. So mentioning it here makes it seem like there's something different about this with Jupyter, but there isn't.

> So maybe the change here is to say:
> > All subsequent method calls will update the state of your :py:class:`.ProtocolContext`, unless you reset it by calling :py:meth:`~opentrons.execute.get_protocol_api` again.

I think that's moving in the wrong direction, because calling `get_protocol_api()` will *not* reset  the `ProtocolContext` that you already have. It will return a *new* `ProtocolContext`, with a blank slate, which we expect you to start using instead of your old `ProtocolContext`. You *could* do things wrong and continue to use the old `ProtocolContext` after getting a new one from `get_protocol_api()`; I don't know what will happen, but probably nothing good.

Reviewer: Anyway, I'm happy for this to be merged as-is so we get the `systemctl` workaround published.

Author: Stripped this paragraph down quite a bit to avoid getting in the weeds.

Running a Previously Written Protocol
+++++++++++++++++++++++++++++++++++++

You can also use Jupyter to run a protocol that you have already written. To do so, first copy the entire text of the protocol into a cell and run that cell:

.. code-block:: python

    import opentrons.execute
    from opentrons import protocol_api
    def run(protocol: protocol_api.ProtocolContext):
        # the contents of your previously written protocol go here

Since a typical protocol only `defines` the ``run`` function but doesn't `call` it, this won't immediately cause the OT-2 to move. To begin the run, instantiate a :py:class:`.ProtocolContext` and pass it to the ``run`` function you just defined:

.. code-block:: python
    :substitutions:

    protocol = opentrons.execute.get_protocol_api('|apiLevel|')
    run(protocol)  # your protocol will now run

Using Custom Labware
++++++++++++++++++++

If you have custom labware definitions you want to use with Jupyter, make a new directory called ``labware`` in Jupyter and put the definitions there. These definitions will be available when you call :py:meth:`~opentrons.protocol_api.contexts.ProtocolContext.load_labware`.

> Using Modules

Reviewer: ```suggestion
Using Hardware Modules
```

To avoid confusion with Python `import` modules.

Author: The "Hardware Modules" page is linked as such in the first line of this section, and lengthening the section name causes some ugly wrap in the sidebar nav. I think this is OK as is.

Custom Labware
++++++++++++++

> If you have custom labware definitions you want to use with Jupyter, make a new directory called ``labware`` in Jupyter and put the definitions there. These definitions will be available when you call ``load_labware``.

Reviewer 2: I think we want to keep the previous quotes for 'labware'. It's not referring to the labware object in code.

Author: I changed it to monospace because it's a file path. Let me check if we have precedent for paths elsewhere.

Author: Looks like we use mono for paths [elsewhere in the Python docs](https://github.com/Opentrons/opentrons/blame/edge/api/docs/v2/writing.rst#L286) and [in some support articles](https://support.opentrons.com/s/article/Using-the-OT-2-s-camera).

Using Modules
+++++++++++++

> If your protocol uses :ref:`new_modules`, you need to take additional steps to make sure that Jupyter Notebook doesn't send commands that conflict with the robot server. If you send commands to modules while the robot server is running, you will likely see errors in Jupyter Notebook and the module commands may not execute as expected.

Reviewer 2: ```suggestion
If your protocol uses :ref:`new_modules`, you need to take additional steps to make sure that Jupyter Notebook doesn't send commands that conflict with the robot server. If you send commands to modules while the robot server is running, you will likely run into errors and the module commands may not execute as expected.
```
(The errors could happen outside of Jupyter too)

Author: Good point, I managed to trigger the error banner in the app too.

Author: Incorporated with slightly different wording.
